### PR TITLE
IBP 3449 Implement BRAPI Get Studies

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 1]
```diff
@@ -786,8 +789,8 @@ public StudyMetadata getStudyMetadataForGeolocationId(final Integer geolocationI
 				studyMetadata.setTrialDbId(
 					(row[5] instanceof Integer) ? (Integer) row[5] : null);
 				studyMetadata.setTrialName((row[6] instanceof String) ? (String) row[6] : null);
-				studyMetadata.setStartDate((row[7] instanceof String) ? (String) row[7] : null);
-				studyMetadata.setEndDate((row[8] instanceof String) ? (String) row[8] : null);
+				studyMetadata.setStartDate(Util.tryParseDate((String) row[7]));
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Is it safe to use `Util.tryParseDate` if there is no date string returned from db?

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 2]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+			sql.append("AND cvtermSeason.cvterm_id = :seasonDbId ");
+		}
+		if (!StringUtils.isEmpty(studySearchFilter.getStudyTypeDbId())) {
+			sql.append("AND pmain.study_type_id = :studyTypeDbId ");
+		}
+		if (!StringUtils.isEmpty(studySearchFilter.getTrialDbId())) {
+			sql.append("AND pmain.project_id = :trialDbId ");
+		}
+		if (studySearchFilter.getActive() != null) {
+			sql.append("AND pmain.deleted = :active ");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Shouldn't we negate `active` value when matching to `deleted` flag?

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 3]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+	private void appendStudySummaryFromQuery(final StringBuilder sql) {
+		sql.append(" FROM ");
+		sql.append("     nd_geolocation geoloc ");
+		sql.append("         INNER JOIN ");
+		sql.append("     nd_experiment nde ON nde.nd_geolocation_id = geoloc.nd_geolocation_id ");
+		sql.append("         INNER JOIN ");
+		sql.append("     project proj ON proj.project_id = nde.project_id ");
+		sql.append("         INNER JOIN ");
+		sql.append("     project pmain ON pmain.project_id = proj.study_id ");
+		sql.append("         LEFT OUTER JOIN ");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: It seems this join is unnecessary given you have separate joins for the season and location geolocationprop records: 
`LEFT OUTER JOIN nd_geolocationprop geoprop `

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 4]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+		sql.append("     nd_geolocationprop geoprop ON geoprop.nd_geolocation_id = geoloc.nd_geolocation_id ");
+		sql.append("         LEFT OUTER JOIN ");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Why do we also need this join to projectprop:
`LEFT OUTER JOIN projectprop pProp`

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 5]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+		sql.append(
+			"     nd_geolocationprop geopropSeason ON geopropSeason.nd_geolocation_id = geoloc.nd_geolocation_id AND geopropSeason.type_id = "
+				+ TermId.SEASON_VAR.getId());
+		sql.append("         LEFT OUTER JOIN ");
+		sql.append(
+			"     nd_geolocationprop geopropLocation ON geopropLocation.nd_geolocation_id = geoloc.nd_geolocation_id AND geopropLocation.type_id = "
+				+ TermId.LOCATION_ID.getId());
+		sql.append("         LEFT OUTER JOIN ");
+		sql.append("     location ON geopropLocation.value = location.locid");
+		sql.append("         LEFT OUTER JOIN ");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: There's no need to join to `workbench.workbench_project.` You can filter by program_uuid using `pmain.program_uuid`

Comment (darla-leafnode) [reply src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Oh nvm, I saw that we need to display program name

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 6]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+		sql.append("     MAX(pmain.end_date) AS " + StudySearchFilter.END_DATE + ", ");
+		sql.append("     CASE WHEN pmain.deleted = 0 THEN 1 ELSE 0 END AS " + StudySearchFilter.ACTIVE + ", ");
+		sql.append("     location.locid AS " + StudySearchFilter.LOCATION_DB_ID + ", ");
+		sql.append("     location.lname AS " + StudySearchFilter.LOCATION_NAME + ", ");
+		sql.append("     wp.project_name AS " + StudySearchFilter.PROGRAM_NAME + ", ");
+		sql.append("     wp.project_uuid AS " + StudySearchFilter.PROGRAM_DB_ID + " ");
+		this.appendStudySummaryFromQuery(sql);
+		this.appendStudySearchFilter(sql, studySearchFilter);
+		sql.append(" GROUP BY geoloc.nd_geolocation_id ");
+		if (!StringUtils.isEmpty(studySearchFilter.getSortedRequest().getSortBy()) && StudySearchFilter.SORTABLE_FIELDS
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Extract these to local variables for better readability:
```
studySearchFilter.getSortedRequest().getSortBy()
studySearchFilter.getSortedRequest().getSortOrder()
```

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java [hunk 7]
```diff
@@ -1400,6 +1403,201 @@ public DatasetDTO getDatasetOfSampleList(final Integer sampleListId) {
...
+			studyDto.setStartDate(Util.tryParseDate((String) result.get(StudySearchFilter.START_DATE)));
+			studyDto.setEndDate(Util.tryParseDate((String) result.get(StudySearchFilter.END_DATE)));
+			studyDto.setLocationDbId(String.valueOf(result.get(StudySearchFilter.LOCATION_DB_ID)));
+			studyDto.setLocationName(String.valueOf(result.get(StudySearchFilter.LOCATION_NAME)));
+			studyDto.setProgramDbId(String.valueOf(result.get(StudySearchFilter.PROGRAM_DB_ID)));
+			studyDto.setProgramName(String.valueOf(result.get(StudySearchFilter.PROGRAM_NAME)));
+
+			final List<SeasonDto> seasons = new ArrayList<>();
+			if (!StringUtils.isEmpty((String) result.get(StudySearchFilter.SEASON_DB_ID))) {
+				final SeasonDto seasonDto = new SeasonDto();
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: Create a `SeasonDto `constructor with `seasonDbId` and `season `parameters

File: src/main/java/org/generationcp/middleware/service/api/study/StudyService.java [hunk 8]
```diff
@@ -75,4 +75,8 @@
 	Integer getPlotDatasetId(final int studyId);
 
 	Integer getEnvironmentDatasetId(final int studyId);
+
+	List<StudyDto> getStudyDTOs(StudySearchFilter studySearchFilter);
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/api/study/StudyService.java]: Rename to `countStudies` and `getStudies` (omit the "DTO")

File: src/main/java/org/generationcp/middleware/util/Util.java [hunk 9]
```diff
@@ -326,6 +326,29 @@ public static Date parseDate(final String date, final String format) throws Pars
...
 
+	/**
+	 * Returns the date object from the specified format
+	 *
+	 * @return date object or null if it cannot be parsed
+	 */
+	public static Date tryParseDate(final String date, final String format) {
+		try {
+			return Util.parseDate(date, format);
+		} catch (final ParseException | NullPointerException e) {
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/util/Util.java]: Several concerns here:
1. We should not be catching NPEs. 
2. Avoid returning null

Comment (nahuel-soldevilla) [reply src/main/java/org/generationcp/middleware/util/Util.java]: This is mine so I can explain. This is just a convenient method to format the value coming from the db, so we don't have to try/catch every single line where this parsing is needed.

@darla-leafnode @abatac

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java (hunk 1 above)
Comment (nahuel-soldevilla) [reply src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: https://github.com/IntegratedBreedingPlatform/Middleware/pull/764#discussion_r392214326

File: src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java (hunk 4 above)
Comment (abatac) [reply src/main/java/org/generationcp/middleware/dao/dms/DmsProjectDao.java]: This is unnecessary, will remove it.

**Comment by abatac** [comment]
Hi @darla-leafnode, I'm done with the requested changes. Please check.

**Pull request merged**